src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -0,0 +1,283 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parser;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class DeclaredTypeManager
+    {
+        // maps syntax nodes to their declared types
+        // processed nodes found not to have a declared type will have a null value
+        private readonly IDictionary<SyntaxBase, DeclaredTypeAssignment?> declaredTypes = new Dictionary<SyntaxBase, DeclaredTypeAssignment?>();
+
+        private readonly SyntaxHierarchy hierarchy;
+        private readonly ITypeManager typeManager;
+        private readonly IResourceTypeProvider resourceTypeProvider;
+        private readonly IReadOnlyDictionary<SyntaxBase, Symbol> bindings;
+
+        public DeclaredTypeManager(SyntaxHierarchy hierarchy, ITypeManager typeManager, IResourceTypeProvider resourceTypeProvider, IReadOnlyDictionary<SyntaxBase,Symbol> bindings)
+        {
+            this.hierarchy = hierarchy;
+            this.typeManager = typeManager;
+            this.resourceTypeProvider = resourceTypeProvider;
+            this.bindings = bindings;
+        }
+
+        public DeclaredTypeAssignment? GetDeclaredTypeAssignment(SyntaxBase syntax)
+        {
+            if (this.declaredTypes.TryGetValue(syntax, out var assignment))
+            {
+                // syntax node has already been processed
+                // return result as-is (even if null)
+                return assignment;
+            }
+
+            // the node has not been processed
+            // figure out the type
+            var newAssignment = GetTypeAssignment(syntax);
```

Comment:
Using a visitor here didn't really make sense as I'm really going up the tree using the parent pointers rather than going down. `GetTypeAssignment` eventually will call recursively back into `GetDeclaredTypeAssignment`. This makes this class callable without any specific order as it'll basically walk up the parents until it finds a declaration, figure out its declared type, and then walk down transforming the type along the way. (Same as before - just a bit cleaner now that it's separated from the rest of the type stuff.)